### PR TITLE
dlopen should follow platform default behavior

### LINKED ISSUE #11322: dlopen RTLD_LOCAL default not respected on OSX
the documentation and code state that dlopen in julia defaults to RTLD_LOCAL, but on mac, it seems that the defaults is RTLD_GLOBAL per https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man3/dlopen.3.html

implemented by @stevengj in 27fab368 and noted while doing code review on #11307

### LINKED ISSUE #11689: "Symbol not found" errors on particular ccalls on OSX
Font rendering broke on Gadfly on 0.4 some time in the past few weeks, and I was able to reduce it to the following test case.

```julia
const libpangocairo = "/usr/local/lib/libpangocairo-1.0.dylib"
const libpango = "/usr/local/lib/libpango-1.0.dylib"

fontmap = ccall((:pango_cairo_font_map_new, libpangocairo),
                Ptr{Void}, ())
context = ccall((:pango_font_map_create_context, libpango),
                Ptr{Void}, (Ptr{Void},), fontmap)
layout = ccall((:pango_layout_new, libpango),
               Ptr{Void}, (Ptr{Void},), context)

extents = Array(Int32, 4)
ccall((:pango_layout_get_extents, libpango),
      Void, (Ptr{Void}, Ptr{Int32}, Ptr{Int32}),
      layout, extents, C_NULL)
```

This works fine on 0.3, but on 0.4 I get
```
(process:29701): Pango-WARNING **: dlopen(/usr/local/Cellar/pango/1.36.8/lib/pango/1.8.0/modules/pango-basic-coretext.so, 2): Symbol not found: _pango_core_text_font_get_ctfont
  Referenced from: /usr/local/Cellar/pango/1.36.8/lib/pango/1.8.0/modules/pango-basic-coretext.so
  Expected in: flat namespace
 in /usr/local/Cellar/pango/1.36.8/lib/pango/1.8.0/modules/pango-basic-coretext.so

(process:29701): Pango-WARNING **: failed to choose a font, expect ugly output. engine-type='PangoRenderCoreText', script='latin'

(process:29701): Pango-WARNING **: failed to choose a font, expect ugly output. engine-type='PangoRenderCoreText', script='common'
```

I can bisect this and see what changed, but maybe someone knows the answer already and can save me the trouble.

### PR DESCRIPTION
**Pull request by jakebolewski**
For OSX the default behavior is to dlopen with `RTLD_GLOBAL` while
on most other POSIX platforms the default is `RTLD_LOCAL`.  This reverts
the behavior introduced in #11352 which made `RTLD_LOCAL` the default for
all platforms which caused issues when loading shared libraries on OSX.

This PR also cleans up some `Libdl` behavior and adds tests for the
`Libdl` module.  `Libdl.dlopen` now prints a detailed error message in
the thrown exception rather than printing to STDERR.  `Libdl.dlsym` now
throws an exception instead of printing an error message to STDERR if
the symbol is not found in the shared library.  `RTLD_*` enum values now
start at 1 instead of 0 to reflect the fact that `RTLD_LOCAL` is not
universal default behavior for POSIX platforms and needs to be detected on
OSX.

cc: @vtjnash, @poulson

closes #11692
fixes #11689
fixes #11322

### CODE REVIEW

File: test/libdl.jl [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+let dl = Libdl.dlopen_e("./foo")
+    @test dl == C_NULL
+end
+
+# test dlpath
+let dl = Libdl.dlopen("./libccalltest")
+    try
+        @test dl != C_NULL
+        @test (abspath(Libdl.dlpath(dl)) ==
```
Comment (tkelman) [inline test/libdl.jl]: use `Base.samefile` or this will fail on case insensitive filesystems

Comment (jakebolewski) [inline test/libdl.jl]: Fixed, thanks for catching that.

**Comment by vtjnash** [comment]
Lgtm. Looks like the tests (for Travis Linux) need some correction however.

**Comment by tkelman** [comment]
so what wound up being different on travis that was leading to failures earlier?

**Comment by jakebolewski** [comment]
a missing comment block

**Pull request merged**